packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +400,340 @@ contract AmmPool is IBlockReceiver {
         override
         online
         onlyExchangeOwner
-        returns (uint numTransactionsConsumed)
+        returns (uint)
     {
-        AuxiliaryData memory auxData = abi.decode(auxiliaryData, (AuxiliaryData));
-        bytes32 exchangeDomainSeparator = exchange.getDomainSeparator();
+        PoolTransaction[] memory poolTransactions = abi.decode(auxiliaryData, (PoolTransaction[]));
+
+        // Cache the domain seperator to save on SLOADs each time it is accessed.
+        Context memory ctx = Context({
+            _block: _block,
+            txIdx: txIdx,
+            DOMAIN_SEPARATOR: DOMAIN_SEPARATOR,
+            exchangeDomainSeparator: exchange.getDomainSeparator(),
+            ammBalancesInAccount: new uint[](tokens.length),
```

Comment:
ammBalancesInAccount => ammLayer2Balances?
ammBalances => ammLayer1Balances?

I'm not sure I understand the meaning of these variables correctly.

Reply:
`ammBalancesInAccount` is the balance in the account directly. Transfers update the balance of the account immediately, deposits/withdrawals don't update the balance immediately because those operations are batched and only done at the end. So `ammBalances` is the virtual balance if all operations would immediately update the balance in the account (which are the ones we need to do the calculations on), `ammBalancesInAccount` is the actual balance in the account so we can keep track of how much we need to deposit/withdraw at the end.